docs/testing/e2e-plan.md

```diff
@@ -0,0 +1,79 @@
+
+Tests have been written that cover the installation and running of the CP packages controller and the [hello-eks-anywhere test application](https://github.com/aws/eks-anywhere-build-tooling/blob/61628f1669044a39ff561385c7032e3b6c7d12ed/projects/aws-containers/hello-eks-anywhere/conf/hello.sh) across the three dimensions mentioned above (EKS-A provider, node OS, and K8s version). Additional tests have been written to exercise Harbor and MetalLB package installations. The results of these tests are reported in the appropriate EKS-A communications channels.
+
+*But the existing tests are not nearly enough to provide confidence that our packages are bug-free.* The current CP E2E framework does not (without significant effort by the test writer) run the existing tests against the versions of their artifacts that matter most[^1]. Nor do they run them for each submitted pull request (PR). These things are not the fault of the tests themselves, nor the tests' writers, but rather they've exposed the limitations of the current CP E2E framework.
+
+[^1]: Early E2E tests were run against production artifacts, generally because they were the only reliably accessible artifacts at the time.
+
+## The End-Goal
+
+To be able to provide confidence that the CP offerings are working across our supported EKS-A versions, EKS-A providers, cluster node OSes, and K8s versions, the current framework must be improved in the following ways:
+
++ **Artifact version control:**
+  Tests should be run against an arbitrary versions of artifacts, defaulting to their latest versions.
+
++ **Timing:**
+  E2E tests should be run on every revision of an "ok to test" PR. This may present some challenges for the current E2E test architecture, as it might require communication between separate AWS pipelines, or failing that, custom code that can communicate with other AWS resources to identify and list available artifacts built in other pipelines.
+
++ **Platform configuration:**
+  A test should be able to exclude specific EKS-A versions, EKS-A providers, cluster node OSes, or K8s versions with which they are incompatible or unsupported. A test will be run on every permutation by default.
+
++ **Configuration review & control:**
+  The test dimensions should be easily viewed, verified, and edited by humans (or reasonable facsimiles thereof). This provides a means of quickly excluding tests categorically for reasons such as hardware outage.
+
+Additionally, it is foreseen that the number of E2E tests that need to be run for each PR will quickly grow, and so it is recommended that these performance goals also be considered:
+
++ **Reusable clusters:** The amount of time required to create and destroy clusters before and after each test will quickly become overwhelming with reasonable amounts of hardware.
```

Comment:
Is there any merit to using the remote cluster functionality of Prow here? In that case we could potentially have long running clusters on each provider, and leverage prow for this? https://github.com/kubernetes/test-infra/blob/master/config/prow/config.yaml#L41   and can overide the job configuration here. https://github.com/kubernetes/test-infra/blob/master/config/jobs/image-pushing/k8s-staging-cloud-provider-aws.yaml#L5

Reply:
> Is there any merit to using the remote cluster functionality of Prow here? In that case we could potentially have long running clusters on each provider, and leverage prow for this? https://github.com/kubernetes/test-infra/blob/master/config/prow/config.yaml#L41 and can overide the job configuration here. https://github.com/kubernetes/test-infra/blob/master/config/jobs/image-pushing/k8s-staging-cloud-provider-aws.yaml#L5

I had initially had this as something I was hoping we could do, but I was speaking with someone (and geez, I thought it was you) whom said that since our tests were triggered via code builder that they wouldn't be able to run on our prow clusters? I wonder with whom I was having that conversation now?

In any event, prow was one of the first things I wanted to leverage to do all of this. After all, it was built for k8s testing. There could possibly be some issues around having clusters that meet all of the various requirements (bare metal, vs vsphere vs docker etc) all accessible for prow to be able to trigger the jobs, but in general, I like the idea of doing it that way, because it feels like a sort of dog fooding. We build on kubernetes, we run on kubernets, we test on kubernetes, that sort of thing. Plus prow might be a candidate for a package itself one day, and then it would be cool to run tests on a prow instance that was itself setup and run by EKS-A. :D 

There are still questions about the suitability of prow though (at least in my mind), because as Binyam pointed out, we need to be testing first and foremost through eksctl-anywhere. I'm not sure how tests are launched on a prow installation, but if we can't launch our jobs via eksctl-anywhere, then that severely limits its usefulness to us as a test-running platform. I'd be surprised if we couldn't, but it's still an unknown (to me).